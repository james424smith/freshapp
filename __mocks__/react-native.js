import * as ReactNative from "react-native";

export * from "react-native";
export const Vibration = { vibrate: jest.fn() };
export const Adnimated = { spring: jest.fn() };

export default Object.setPrototypeOf(
  {
    Vibration,
    Adnimated,
  },
  ReactNative
);
