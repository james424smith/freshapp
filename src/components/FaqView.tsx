import React, { useRef } from "react";
import { FlatList, Linking, View } from "react-native";
import { Icon } from "react-native-elements";
import { TouchableOpacity } from "react-native-gesture-handler";
import styles from "../styles/FAQStyles";
import Text from "./StyledText";
import colors from "../constants/styles/colors";
import { useTheme } from "@react-navigation/native";
import { useSelectedQuestion } from "../hooks/useSelectedQuestion";
import { heightPercentageToDP as hp } from "react-native-responsive-screen";

interface Props {
  content: {
    question: string;
    answer: string;
    opened: boolean;
    id: number;
    url?: string;
  }[];
  selectedQuestion?: number;
}

const FaqView = (props: Props) => {
  const theme = useTheme();
  const flatlistRef = useRef<FlatList<{
    question: string;
    answer: string;
    opened: boolean;
    id: number;
    url?: string;
  }> | null>(null);

  const { answerAndquestions, setAnswerAndquestions } = useSelectedQuestion(
    props.content,
    flatlistRef,
    props.selectedQuestion
  );

  const setOpenedQuestion = (index: any) => {
    const temp = answerAndquestions.map((item, arrayIndex) => {
      if (index === arrayIndex) {
        return { ...item, opened: !item.opened };
      } else {
        return { ...item, opened: false };
      }
    });
    setAnswerAndquestions(temp);
  };

  const _renderAnswerAndQuestions = ({
    item,
    index,
  }: {
    item: {
      question: string;
      answer: string;
      opened: boolean;
      id: number;
      url?: string;
    };
    index: number;
  }) => {
    return (
      <TouchableOpacity
        testID={`faq-button-${item.id}`}
        style={styles.questionPress}
        onPress={() => {
          setOpenedQuestion(index);
        }}
      >
        <View style={styles.questionView}>
          <Text
            style={{
              ...styles.questionText,
              color: theme.colors.FAQQuestionToDark,
            }}
          >
            {item.question}
          </Text>
          <Icon
            name={item.opened ? "keyboard-arrow-up" : "keyboard-arrow-down"}
            color={colors.darkInputText}
            size={30}
          />
        </View>
        {item.opened ? (
          <TouchableOpacity
            testID={"open-url-button"}
            onPress={() => {
              item.url && Linking.openURL(item.url);
            }}
            disabled={!item.url}
            style={styles.answerView}
          >
            <Text
              style={{
                ...styles.answerText,
                color: theme.colors.FAQQuestionToDark,
              }}
            >
              {item.answer}
            </Text>
          </TouchableOpacity>
        ) : null}
      </TouchableOpacity>
    );
  };

  return (
    <FlatList
      ref={flatlistRef}
      getItemLayout={(_data, index) => ({
        length: hp("3%"),
        offset: hp("3%") * index,
        index,
      })}
      data={answerAndquestions}
      renderItem={_renderAnswerAndQuestions}
      keyExtractor={(_item, i) => i.toString()}
    />
  );
};
export default FaqView;
