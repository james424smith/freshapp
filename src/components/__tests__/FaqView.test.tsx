import React from "react";
import FaqView from "../FaqView";
import { render, cleanup, fireEvent } from "@testing-library/react-native";
import { Linking } from "react-native";

describe("EmployeeIdInput snapshot test", () => {
  afterEach(() => {
    cleanup();
    jest.clearAllMocks();
    jest.restoreAllMocks();
  });

  it("should render correct ui and change value of employeeId", () => {
    const { toJSON, getByTestId } = render(
      <FaqView
        content={[
          {
            question: "Some Question",
            answer: "Some Answer",
            opened: true,
            id: 1,
          },
          {
            question: "Some Question 2",
            answer: "Some Answer 2",
            opened: false,
            id: 2,
          },
        ]}
        selectedQuestion={1}
      />
    );

    fireEvent.press(getByTestId("faq-button-2"));

    expect(toJSON()).toMatchSnapshot();
  });

  it("should render correct ui and change value of employeeId whem selectedQuestion is not present in the questions array", () => {
    const { toJSON } = render(
      <FaqView
        content={[
          {
            question: "Some Question",
            answer: "Some Answer",
            opened: true,
            id: 1,
          },
        ]}
        selectedQuestion={2}
      />
    );

    expect(toJSON()).toMatchSnapshot();
  });

  it("should call the openUrl function when clicking an item with url property", () => {
    const { getByTestId } = render(
      <FaqView
        content={[
          {
            question: "Some Question",
            answer: "Some Answer",
            opened: true,
            url: "https://www.google.com",
            id: 1,
          },
        ]}
      />
    );

    Linking.openURL = jest.fn();

    fireEvent.press(getByTestId("open-url-button"));

    expect(Linking.openURL).toHaveBeenCalledWith("https://www.google.com");
  });
});
