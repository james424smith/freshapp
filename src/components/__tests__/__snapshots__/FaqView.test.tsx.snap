// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EmployeeIdInput snapshot test should render correct ui and change value of employeeId 1`] = `
<RCTScrollView
  data={
    Array [
      Object {
        "answer": "Some Answer",
        "id": 1,
        "opened": false,
        "question": "Some Question",
      },
      Object {
        "answer": "Some Answer 2",
        "id": 2,
        "opened": true,
        "question": "Some Question 2",
      },
    ]
  }
  disableVirtualization={false}
  getItem={[Function]}
  getItemCount={[Function]}
  getItemLayout={[Function]}
  horizontal={false}
  initialNumToRender={10}
  keyExtractor={[Function]}
  maxToRenderPerBatch={10}
  onContentSizeChange={[Function]}
  onEndReachedThreshold={2}
  onLayout={[Function]}
  onMomentumScrollEnd={[Function]}
  onScroll={[Function]}
  onScrollBeginDrag={[Function]}
  onScrollEndDrag={[Function]}
  removeClippedSubviews={false}
  renderItem={[Function]}
  scrollEventThrottle={50}
  stickyHeaderIndices={Array []}
  updateCellsBatchingPeriod={50}
  viewabilityConfigCallbackPairs={Array []}
  windowSize={21}
>
  <View>
    <View
      style={null}
    >
      <RNGestureHandlerButton
        collapsable={false}
        onGestureEvent={[Function]}
        onGestureHandlerEvent={[Function]}
        onGestureHandlerStateChange={[Function]}
        onHandlerStateChange={[Function]}
        rippleColor={0}
        testID="faq-button-1"
      >
        <View
          accessible={true}
          style={
            Object {
              "borderBottomColor": "#C5C5C5",
              "borderBottomWidth": 0.3,
              "opacity": 1,
              "paddingVertical": 20,
            }
          }
        >
          <View
            style={
              Object {
                "flexDirection": "row",
                "justifyContent": "space-between",
                "marginLeft": 22.5,
                "marginRight": 30,
              }
            }
          >
            <Text
              style={
                Array [
                  Object {
                    "color": "#2A2E43",
                    "fontFamily": "Roboto-Regular",
                  },
                  Object {
                    "color": undefined,
                    "fontSize": 14,
                    "fontWeight": "bold",
                    "width": 600,
                  },
                ]
              }
            >
              Some Question
            </Text>
            <View
              style={
                Object {
                  "overflow": "hidden",
                }
              }
            >
              <View>
                <View
                  style={
                    Object {
                      "alignItems": "center",
                      "backgroundColor": "transparent",
                      "justifyContent": "center",
                    }
                  }
                >
                  <Text
                    allowFontScaling={false}
                    style={
                      Array [
                        Object {
                          "color": "#8E9397",
                          "fontSize": 30,
                        },
                        Object {
                          "backgroundColor": "transparent",
                        },
                        Object {
                          "fontFamily": "Material Icons",
                          "fontStyle": "normal",
                          "fontWeight": "normal",
                        },
                        Object {},
                      ]
                    }
                    testID="iconIcon"
                  >
                    
                  </Text>
                </View>
              </View>
            </View>
          </View>
        </View>
      </RNGestureHandlerButton>
    </View>
    <View
      style={null}
    >
      <RNGestureHandlerButton
        collapsable={false}
        onGestureEvent={[Function]}
        onGestureHandlerEvent={[Function]}
        onGestureHandlerStateChange={[Function]}
        onHandlerStateChange={[Function]}
        rippleColor={0}
        testID="faq-button-2"
      >
        <View
          accessible={true}
          style={
            Object {
              "borderBottomColor": "#C5C5C5",
              "borderBottomWidth": 0.3,
              "opacity": 1,
              "paddingVertical": 20,
            }
          }
        >
          <View
            style={
              Object {
                "flexDirection": "row",
                "justifyContent": "space-between",
                "marginLeft": 22.5,
                "marginRight": 30,
              }
            }
          >
            <Text
              style={
                Array [
                  Object {
                    "color": "#2A2E43",
                    "fontFamily": "Roboto-Regular",
                  },
                  Object {
                    "color": undefined,
                    "fontSize": 14,
                    "fontWeight": "bold",
                    "width": 600,
                  },
                ]
              }
            >
              Some Question 2
            </Text>
            <View
              style={
                Object {
                  "overflow": "hidden",
                }
              }
            >
              <View>
                <View
                  style={
                    Object {
                      "alignItems": "center",
                      "backgroundColor": "transparent",
                      "justifyContent": "center",
                    }
                  }
                >
                  <Text
                    allowFontScaling={false}
                    style={
                      Array [
                        Object {
                          "color": "#8E9397",
                          "fontSize": 30,
                        },
                        Object {
                          "backgroundColor": "transparent",
                        },
                        Object {
                          "fontFamily": "Material Icons",
                          "fontStyle": "normal",
                          "fontWeight": "normal",
                        },
                        Object {},
                      ]
                    }
                    testID="iconIcon"
                  >
                    
                  </Text>
                </View>
              </View>
            </View>
          </View>
          <RNGestureHandlerButton
            collapsable={false}
            onGestureEvent={[Function]}
            onGestureHandlerEvent={[Function]}
            onGestureHandlerStateChange={[Function]}
            onHandlerStateChange={[Function]}
            rippleColor={0}
            testID="open-url-button"
          >
            <View
              accessible={true}
              style={
                Object {
                  "flexDirection": "row",
                  "marginLeft": 22.5,
                  "marginRight": 45,
                  "opacity": 1,
                }
              }
            >
              <Text
                style={
                  Array [
                    Object {
                      "color": "#2A2E43",
                      "fontFamily": "Roboto-Regular",
                    },
                    Object {
                      "color": undefined,
                      "fontSize": 14,
                      "marginRight": 52.5,
                      "paddingTop": 40,
                    },
                  ]
                }
              >
                Some Answer 2
              </Text>
            </View>
          </RNGestureHandlerButton>
        </View>
      </RNGestureHandlerButton>
    </View>
  </View>
</RCTScrollView>
`;

exports[`EmployeeIdInput snapshot test should render correct ui and change value of employeeId whem selectedQuestion is not present in the questions array 1`] = `
<RCTScrollView
  data={
    Array [
      Object {
        "answer": "Some Answer",
        "id": 1,
        "opened": true,
        "question": "Some Question",
      },
    ]
  }
  disableVirtualization={false}
  getItem={[Function]}
  getItemCount={[Function]}
  getItemLayout={[Function]}
  horizontal={false}
  initialNumToRender={10}
  keyExtractor={[Function]}
  maxToRenderPerBatch={10}
  onContentSizeChange={[Function]}
  onEndReachedThreshold={2}
  onLayout={[Function]}
  onMomentumScrollEnd={[Function]}
  onScroll={[Function]}
  onScrollBeginDrag={[Function]}
  onScrollEndDrag={[Function]}
  removeClippedSubviews={false}
  renderItem={[Function]}
  scrollEventThrottle={50}
  stickyHeaderIndices={Array []}
  updateCellsBatchingPeriod={50}
  viewabilityConfigCallbackPairs={Array []}
  windowSize={21}
>
  <View>
    <View
      style={null}
    >
      <RNGestureHandlerButton
        collapsable={false}
        onGestureEvent={[Function]}
        onGestureHandlerEvent={[Function]}
        onGestureHandlerStateChange={[Function]}
        onHandlerStateChange={[Function]}
        rippleColor={0}
        testID="faq-button-1"
      >
        <View
          accessible={true}
          style={
            Object {
              "borderBottomColor": "#C5C5C5",
              "borderBottomWidth": 0.3,
              "opacity": 1,
              "paddingVertical": 20,
            }
          }
        >
          <View
            style={
              Object {
                "flexDirection": "row",
                "justifyContent": "space-between",
                "marginLeft": 22.5,
                "marginRight": 30,
              }
            }
          >
            <Text
              style={
                Array [
                  Object {
                    "color": "#2A2E43",
                    "fontFamily": "Roboto-Regular",
                  },
                  Object {
                    "color": undefined,
                    "fontSize": 14,
                    "fontWeight": "bold",
                    "width": 600,
                  },
                ]
              }
            >
              Some Question
            </Text>
            <View
              style={
                Object {
                  "overflow": "hidden",
                }
              }
            >
              <View>
                <View
                  style={
                    Object {
                      "alignItems": "center",
                      "backgroundColor": "transparent",
                      "justifyContent": "center",
                    }
                  }
                >
                  <Text
                    allowFontScaling={false}
                    style={
                      Array [
                        Object {
                          "color": "#8E9397",
                          "fontSize": 30,
                        },
                        Object {
                          "backgroundColor": "transparent",
                        },
                        Object {
                          "fontFamily": "Material Icons",
                          "fontStyle": "normal",
                          "fontWeight": "normal",
                        },
                        Object {},
                      ]
                    }
                    testID="iconIcon"
                  >
                    
                  </Text>
                </View>
              </View>
            </View>
          </View>
          <RNGestureHandlerButton
            collapsable={false}
            onGestureEvent={[Function]}
            onGestureHandlerEvent={[Function]}
            onGestureHandlerStateChange={[Function]}
            onHandlerStateChange={[Function]}
            rippleColor={0}
            testID="open-url-button"
          >
            <View
              accessible={true}
              style={
                Object {
                  "flexDirection": "row",
                  "marginLeft": 22.5,
                  "marginRight": 45,
                  "opacity": 1,
                }
              }
            >
              <Text
                style={
                  Array [
                    Object {
                      "color": "#2A2E43",
                      "fontFamily": "Roboto-Regular",
                    },
                    Object {
                      "color": undefined,
                      "fontSize": 14,
                      "marginRight": 52.5,
                      "paddingTop": 40,
                    },
                  ]
                }
              >
                Some Answer
              </Text>
            </View>
          </RNGestureHandlerButton>
        </View>
      </RNGestureHandlerButton>
    </View>
  </View>
</RCTScrollView>
`;
